src/librustc/ty/print/pretty.rs

```diff
@@ -969,69 +986,129 @@ pub trait PrettyPrinter<'tcx>:
                     _ => p!(write("{}{}", sign_extend(data, size) as i128, i_str))
                 }
             },
-            (ConstValue::Scalar(Scalar::Raw { data, .. }), ty::Char) =>
+            // Char
+            (Scalar::Raw { data, .. }, ty::Char) =>
                 p!(write("{:?}", ::std::char::from_u32(data as u32).unwrap())),
-            (ConstValue::Scalar(_), ty::RawPtr(_)) => p!(write("{{pointer}}")),
-            (ConstValue::Scalar(Scalar::Ptr(ptr)), ty::FnPtr(_)) => {
+            // References and pointers
+            (Scalar::Raw { data: 0, .. }, ty::RawPtr(_)) => p!(write("{{null pointer}}")),
+            // This is UB, but we still print it
+            (Scalar::Raw { data: 0, ..}, ty::Ref(_, ty, _))
+                => p!(write("{{null reference to "), print(ty), write("}}")),
+            (Scalar::Raw { data, .. }, ty::Ref(..)) |
+            (Scalar::Raw { data, .. }, ty::RawPtr(_)) => {
+                let pointer_width = self.tcx().data_layout.pointer_size.bytes();
+                p!(write("0x{:01$x}", data, pointer_width as usize * 2))
+            },
+            (Scalar::Ptr(p), ty::Ref(..)) |
+            (Scalar::Ptr(p), ty::RawPtr(_)) => self = self.pretty_print_const_pointer(p)?,
```

Comment:
Please put the part after the `=>` on its own line, it otherwise gets jumped over visually too easily.